From 5ebd4e4d5a41a18ef894a4cc3c275259085552b7 Mon Sep 17 00:00:00 2001
From: TurtleRuss <tttturtleruss@gmail.com>
Date: Thu, 28 Nov 2024 10:18:16 +0800
Subject: [PATCH 04/10] chore: fix unused varible


diff --git a/SVF/svf/lib/MSSA/MemSSA.cpp b/SVF/svf/lib/MSSA/MemSSA.cpp
index 6831d72ed..1245a557b 100644
--- a/SVF/svf/lib/MSSA/MemSSA.cpp
+++ b/SVF/svf/lib/MSSA/MemSSA.cpp
@@ -666,7 +666,7 @@ void MemSSA::dumpMSSA(OutStream& Out)
                         }
                     }
 
-                    if (has_debug_info)
+                    if (has_mu_or_chi && has_debug_info)
                         Out << "SourceLoc->" << inst->getSourceLoc() << "\n";
 
                     if (has_chi)
-- 
2.39.5 (Apple Git-154)

