From f8036200381b2b3dd6e370ca1ce25ccff02cb0b2 Mon Sep 17 00:00:00 2001
From: TurtleRuss <tttturtleruss@gmail.com>
Date: Tue, 4 Mar 2025 19:18:33 +0800
Subject: [PATCH 09/10] fix: dump source loc of CallInst


diff --git a/SVF/svf/include/MSSA/MSSAMuChi.h b/SVF/svf/include/MSSA/MSSAMuChi.h
index 96e9e7c77..9ee98a6fa 100644
--- a/SVF/svf/include/MSSA/MSSAMuChi.h
+++ b/SVF/svf/include/MSSA/MSSAMuChi.h
@@ -265,7 +265,9 @@ public:
     virtual void dump()
     {
         SVFUtil::outs() << "CALMU(MR_" << this->getMR()->getMRID() << "V_" << this->getMRVer()->getSSAVersion() << ") \t" <<
-                        this->getMR()->dumpStr() << "\n" << this->getCallSite()->getSourceLoc() << '\n';
+                        this->getMR()->dumpStr() << "\n";
+        SVFUtil::outs() << "SourceLoc->" << this->getCallSite()->getSourceLoc() << '\n'
+                        << "FunctionLoc->" << this->getCallSite()->getFun()->getSourceLoc() << '\n';
     }
 };
 
@@ -567,7 +569,9 @@ public:
     {
         SVFUtil::outs() << this->getMR()->getMRID() << "V_" << this->getResVer()->getSSAVersion() <<
                         " = CALCHI(MR_" << this->getMR()->getMRID() << "V_" << this->getOpVer()->getSSAVersion() << ") \t" <<
-                        this->getMR()->dumpStr() << "\n" << this->getCallSite()->getSourceLoc() << '\n';
+                        this->getMR()->dumpStr() << "\n";
+        SVFUtil::outs() << "SourceLoc->" << this->getCallSite()->getSourceLoc() << '\n'
+                        << "FunctionLoc->" << this->getCallSite()->getFun()->getSourceLoc() << '\n';
     }
 };
 
diff --git a/SVF/svf/lib/MSSA/MemSSA.cpp b/SVF/svf/lib/MSSA/MemSSA.cpp
index 3a862b48a..ec6edff8a 100644
--- a/SVF/svf/lib/MSSA/MemSSA.cpp
+++ b/SVF/svf/lib/MSSA/MemSSA.cpp
@@ -727,8 +727,10 @@ void MemSSA::dumpMSSA(OutStream& Out)
                         }
                     }
 
-                    if (has_mu_or_chi && has_debug_info)
+                    if (has_mu_or_chi && has_debug_info) {
                         Out << "SourceLoc->" << inst->getSourceLoc() << "\n";
+                        Out << "FunctionLoc->" << inst->getFun()->getSourceLoc() << '\n';
+                    }
 
                     if (has_chi)
                     {
-- 
2.39.5 (Apple Git-154)

